### PR TITLE
[WIP] add pragma statements

### PR DESCRIPTION
**Pull request by scharlottej13**
- [ ] partially addresses #5445, specifically:

> Pragma no cover
> - We have cases with relatively long switch-like statements ending in a final exception, e.g. ValueError "unknown value"
> - Conditional imports may need be covered with pragma statements and/or a dedicated job

WIP, because I haven't made it through all the files yet. Going in order of lowest to highest percent [coverage](https://app.codecov.io/gh/dask/distributed) I made it to `utils.py`.

cc: @ncclementi @phobson-- thanks again for your help!

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by scharlottej13** [comment]
Some specific questions I had going through this:

[node.py](https://app.codecov.io/gh/dask/distributed/blob/main/distributed/node.py)-- wasn't sure about [`start_services`](https://github.com/dask/distributed/blob/713204b9757065cc1ce6df3574957a79a9f56ea6/distributed/node.py#L31)

[utils_test.py](https://app.codecov.io/gh/dask/distributed/blob/main/distributed/utils_test.py)-- This one was tricky and perhaps should be removed from this PR to be part of a separate discussion (per the note in the original issue)? When unsure, I added `pragma no cover` where the only reference was in a test that is already ignored in `.coveragerc`; if the function was called elsewhere in `utils_test.py`, I opted to not exclude it. For some functions, I had trouble finding a reference, e.g [`deep`](https://github.com/dask/distributed/blob/713204b9757065cc1ce6df3574957a79a9f56ea6/distributed/utils_test.py#L291), [`slowdouble`](https://github.com/dask/distributed/blob/713204b9757065cc1ce6df3574957a79a9f56ea6/distributed/utils_test.py#L316).

[profile.py](https://app.codecov.io/gh/dask/distributed/blob/main/distributed/profile.py)
- [line 110](https://github.com/dask/distributed/blob/27e467e56e1dea6a186152551a269de180bb593e/distributed/profile.py#L110)
- [line 368](https://github.com/dask/distributed/blob/27e467e56e1dea6a186152551a269de180bb593e/distributed/profile.py#L368)
- [lines 444-503](https://github.com/dask/distributed/blob/27e467e56e1dea6a186152551a269de180bb593e/distributed/profile.py#L444-L503) (how hard is it to test these?)

[_concurrent_futures_thread.py](https://app.codecov.io/gh/dask/distributed/blob/main/distributed/_concurrent_futures_thread.py)-- really not sure about this one! seems a lot is not covered, and I wasn't sure how easy this is to test.

**Comment by fjetter** [comment]
Thank you for your effort on this! I would actually recommend opening the other files in another PR, doing this incrementally in batches. That makes reviewing a bit easier and we can merge incrementally to avoid merge conflicts

> [node.py](https://app.codecov.io/gh/dask/distributed/blob/main/distributed/node.py)-- wasn't sure about [start_services](https://github.com/dask/distributed/blob/713204b9757065cc1ce6df3574957a79a9f56ea6/distributed/node.py#L31)

I think this is a good example for a thing we do not want to ignore. We are clearly entering this function (the `if` and `for` is hit) but for some reason the actual logic is never executed. We likely want to spend some time figuring out why this is and maybe wrote a test for it.

> his one was tricky and perhaps should be removed from this PR to be part of a separate discussion

If you are not sure about this, feel free to split it off into a separate PR. In my mind the initial PRs for this should be fast and without a lot of discussion

> [line 110](https://github.com/dask/distributed/blob/27e467e56e1dea6a186152551a269de180bb593e/distributed/profile.py#L110)
> [line 368](https://github.com/dask/distributed/blob/27e467e56e1dea6a186152551a269de180bb593e/distributed/profile.py#L368)

I suggest to not ignore. I don't know how this is tested right now but both lines look like actual functionality which we _should_ be testing imo

> [lines 444-503](https://github.com/dask/distributed/blob/27e467e56e1dea6a186152551a269de180bb593e/distributed/profile.py#L444-L503) (how hard is it to test these?)

They are hit from what I see. If you have reports were they are not hit, this is one of these "flaky" coverage places where it _might_ pay off to have a closer look so I suggest to not ignore

File: distributed/deploy/spec.py [hunk 1]
```diff
@@ -271,7 +271,7 @@ async def _start(self):
         self._lock = asyncio.Lock()
         self.status = Status.starting
 
-        if self.scheduler_spec is None:
```
Comment (fjetter) [inline distributed/deploy/spec.py:274]: keep this, please. This smells like dead code or something we _should_ be testing

File: distributed/profile.py [hunk 2]
```diff
@@ -348,7 +348,7 @@ def get_profile(history, recent=None, start=None, stop=None, key=None):
 
     if stop is None:
         istop = None
-    else:
```
Comment (fjetter) [inline distributed/profile.py:351]: keep this please. Either it should be tested or it shouldn't be there.

File: distributed/utils_test.py [hunk 3]
```diff
@@ -29,15 +29,15 @@
...
-if TYPE_CHECKING:
+if TYPE_CHECKING:  # pragma: no cover
     from typing_extensions import Literal
 
 from distributed.compatibility import MACOS
 from distributed.scheduler import Scheduler
 
 try:
     import ssl
-except ImportError:
```
Comment (fjetter) [inline distributed/utils_test.py:40]: we could consider adding except ImportError as well to the exclude_lines

**Review by fjetter — COMMENTED** [review]
I agree that we should split off the `utils.py` to another PR. I think a lot of these ignores should not be there and this may be an opportunity to clean up our API a little bit

**Comment by scharlottej13** [comment]
> Thank you for your effort on this! I would actually recommend opening the other files in another PR, doing this incrementally in batches. That makes reviewing a bit easier and we can merge incrementally to avoid merge conflicts

Thank you @fjetter for the review! Once I address your comments I'll open up separate PRs for the other files, starting with `utils_test.py`.

> 
> > [lines 444-503](https://github.com/dask/distributed/blob/27e467e56e1dea6a186152551a269de180bb593e/distributed/profile.py#L444-L503) (how hard is it to test these?)
> 
> They are hit from what I see. If you have reports were they are not hit, this is one of these "flaky" coverage places where it _might_ pay off to have a closer look so I suggest to not ignore

Oh interesting, yup in the most recent coverage report I can see they are covered. I won't ignore them, then, and will also see if I can find the report where I saw them marked as not covered.

**Review by fjetter — APPROVED** [review]
Thank you @scharlottej13 !

**Pull request merged**